include/deal.II/lac/petsc_precondition.h

```diff
@@ -1090,6 +1080,67 @@ namespace PETScWrappers
     initialize();
   };
 
+  class PreconditionShell : public PreconditionBase
+  {
+  public:
+    /**
+     * Empty Constructor. You need to call initialize() before using this
+     * object.
+     */
+    PreconditionShell() = default;
+
+    /**
+     * Constructor. Take the matrix which is used to form the preconditioner.
+     */
+    PreconditionShell(const MatrixBase &matrix);
+
+    /**
+     * Same as above but without setting a matrix to form the preconditioner.
+     */
+    PreconditionShell(const MPI_Comm &communicator);
+
+
+    /**
+     * The callback for the application of the preconditioner. Defaults to
+     * a copy operation if not provided.
+     */
+    std::function<int(VectorBase &dst, const VectorBase &src)> apply;
```

Comment:
Can you say what is being copied? I assume it is `dst = src`, which implies the identity matrix as preconditioner. But why is that useful? Why not just throw an exception?

Reply:
@bangerth The rationale for having defaults for `apply` and `applyT` is based on my laziness. I agree with you it can lead to unexpected behavior. I will remove it and raise an error.

Reply:
done